File: ios/MRT/Core/Network/MessageDispatcher.swift
import Foundation

final class MessageDispatcher {
    private(set) var messages: [ChatMessage] = []
    private(set) var sessions: [SessionModel] = []
    private(set) var pendingApproval: Mrt_ApprovalRequest?
    private(set) var state: ConnectionState = .disconnected
    private(set) var agentInfo: Mrt_AgentInfo?

    func apply(_ envelope: Mrt_Envelope) {
        guard case .event(let event)? = envelope.payload else {
            return
        }

        switch event.evt {
        case .codexOutput(let output):
            applyCodexOutput(output)
        case .approvalRequest(let approval):
            pendingApproval = approval
            state = .showingApproval
        case .statusUpdate(let update):
            applyStatus(update.status)
        case .sessionList(let update):
            sessions = update.sessions.map(SessionModel.init)
            if let pendingApproval,
               sessions.contains(where: { $0.id == pendingApproval.sessionID }) == false {
                clearPendingApproval()
            }
        case .agentInfo(let info):
            agentInfo = info
            state = .connected
        case .error(let error):
            messages.append(
                ChatMessage(
                    sessionID: nil,
                    content: error.message,
                    isComplete: true,
                    role: .system
                )
            )
            if error.fatal {
                state = .reconnecting
            } else if state != .disconnected {
                state = .connected
            }
        case .none:
            break
        }
    }

    func clearPendingApproval() {
        pendingApproval = nil
        if state == .showingApproval {
            state = .connected
        }
    }

    private func applyCodexOutput(_ output: Mrt_CodexOutput) {
        if let lastIndex = messages.indices.last,
           messages[lastIndex].sessionID == output.sessionID,
           messages[lastIndex].role == .assistant,
           !messages[lastIndex].isComplete {
            messages[lastIndex].content += output.content
            messages[lastIndex].isComplete = output.isComplete
            return
        }

        messages.append(
            ChatMessage(
                sessionID: output.sessionID,
                content: output.content,
                isComplete: output.isComplete,
                role: .assistant
            )
        )
    }

    private func applyStatus(_ status: Mrt_TaskStatus) {
        switch status {
        case .running:
            state = .loading
        case .waitingApproval:
            state = .showingApproval
        case .completed, .idle, .cancelled, .error:
            state = .connected
        case .unspecified, .UNRECOGNIZED:
            break
        }
    }
}

File: android/app/src/main/java/com/mrt/app/core/network/MessageDispatcher.kt
package com.mrt.app.core.network

import com.mrt.app.core.models.ChatMessage
import com.mrt.app.core.models.SessionModel
import mrt.Mrt

class MessageDispatcher {
    var messages: List<ChatMessage> = emptyList()
        private set
    var sessions: List<SessionModel> = emptyList()
        private set
    var pendingApproval: Mrt.ApprovalRequest? = null
        private set
    var state: ConnectionState = ConnectionState.DISCONNECTED
        private set
    var agentInfo: Mrt.AgentInfo? = null
        private set

    fun apply(envelope: Mrt.Envelope) {
        if (envelope.payloadCase != Mrt.Envelope.PayloadCase.EVENT) {
            return
        }

        when (envelope.event.evtCase) {
            Mrt.AgentEvent.EvtCase.CODEX_OUTPUT -> applyCodexOutput(envelope.event.codexOutput)
            Mrt.AgentEvent.EvtCase.APPROVAL_REQUEST -> {
                pendingApproval = envelope.event.approvalRequest
                state = ConnectionState.SHOWING_APPROVAL
            }
            Mrt.AgentEvent.EvtCase.STATUS_UPDATE -> applyStatus(envelope.event.statusUpdate.status)
            Mrt.AgentEvent.EvtCase.SESSION_LIST ->
                sessions = envelope.event.sessionList.sessionsList.map(SessionModel::fromProto)

            Mrt.AgentEvent.EvtCase.AGENT_INFO -> {
                agentInfo = envelope.event.agentInfo
                state = ConnectionState.CONNECTED
            }
            Mrt.AgentEvent.EvtCase.ERROR -> {
                val error = envelope.event.error
                messages = messages + ChatMessage(
                    sessionId = null,
                    content = error.message,
                    isComplete = true,
                    role = ChatMessage.Role.SYSTEM,
                )
                state = if (error.fatal) {
                    ConnectionState.RECONNECTING
                } else if (state != ConnectionState.DISCONNECTED) {
                    ConnectionState.CONNECTED
                } else {
                    ConnectionState.DISCONNECTED
                }
            }
            Mrt.AgentEvent.EvtCase.EVT_NOT_SET -> Unit
            null -> Unit
        }
    }

    fun clearPendingApproval() {
        pendingApproval = null
        if (state == ConnectionState.SHOWING_APPROVAL) {
            state = ConnectionState.CONNECTED
        }
    }

    private fun applyCodexOutput(output: Mrt.CodexOutput) {
        val current = messages.lastOrNull()
        if (
            current != null &&
            current.sessionId == output.sessionId &&
            current.role == ChatMessage.Role.ASSISTANT &&
            !current.isComplete
        ) {
            messages = messages.dropLast(1) + current.copy(
                content = current.content + output.content,
                isComplete = output.isComplete,
            )
            return
        }

        messages = messages + ChatMessage(
            sessionId = output.sessionId,
            content = output.content,
            isComplete = output.isComplete,
            role = ChatMessage.Role.ASSISTANT,
        )
    }

    private fun applyStatus(status: Mrt.TaskStatus) {
        state = when (status) {
            Mrt.TaskStatus.RUNNING -> ConnectionState.LOADING
            Mrt.TaskStatus.WAITING_APPROVAL -> ConnectionState.SHOWING_APPROVAL
            Mrt.TaskStatus.COMPLETED,
            Mrt.TaskStatus.IDLE,
            Mrt.TaskStatus.CANCELLED,
            Mrt.TaskStatus.ERROR,
            Mrt.TaskStatus.TASK_STATUS_UNSPECIFIED,
            Mrt.TaskStatus.UNRECOGNIZED,
            -> ConnectionState.CONNECTED
        }
    }
}
